source/geod24/Queue.d

```diff
@@ -0,0 +1,391 @@
+
+        Atomically put one element into the queue.
+
+        Params:
+            value = value to add
+
+    ***************************************************************************/
+
+    public void enqueue (T value);
+
+    /***************************************************************************
+
+        Atomically take one element from the queue. Wait blocking or spinning.
+
+        Return:
+            The element at the front of the queue
+
+    ***************************************************************************/
+
+    public T dequeue ();
+
+    /***************************************************************************
+
+        Atomically take one element from the queue.
+
+        Params:
+            value = value to take
+
+        Return:
+            true if a element is retrieved, otherwise false
+
+    ***************************************************************************/
+
+    public bool tryDequeue(out T value);
+}
+
+/*******************************************************************************
+
+    Item of queue
+
+*******************************************************************************/
+
+private static class QueueNode (T)
+{
+    public QueueNode!T next;
+    public T value;
+}
+
+
+/*******************************************************************************
+
+    Blocking queue
+
+*******************************************************************************/
+
+public class BlockingQueue (T) : Queue!T
+{
+    private QueueNode!T head;
+    private QueueNode!T tail;
+    private Mutex head_lock;
+    private Mutex tail_lock;
+
+    /// Ctor
+    public this()
+    {
+        auto n = new QueueNode!T();
+        this.head = this.tail = n;
+        this.head_lock = new Mutex();
+        this.tail_lock = new Mutex();
+    }
+
+    /***************************************************************************
+
+        Atomically put one element into the queue.
+
+        Params:
+            value = value to add
+
+    ***************************************************************************/
+
+    public void enqueue(T value)
+    {
+        auto item = new QueueNode!T();
+        this.tail_lock.lock();
+        scope (exit) this.tail_lock.unlock();
+
+        auto back = this.tail;
+        this.tail = item;
+        back.value = value;
+        atomicFence();
+        back.next = item;
+    }
+
+    /***************************************************************************
+
+        Atomically take one element from the queue. Wait blocking or spinning.
+
+        Return:
+            The element at the front of the queue
+
+    ***************************************************************************/
+
+    public T dequeue ()
+    {
+        this.head_lock.lock();
+        scope (exit) this.head_lock.unlock();
+
+        while (true)
+        {
+            auto head = this.head;
+            auto second = head.next;
+            if (second !is null)
```

Comment:
So if `second` is `null` (there's only one item in the queue) there is an infinite loop ?

Reply:
Ah I went down the code and saw that this is intentional. [Busy waiting](https://en.wikipedia.org/wiki/Busy_waiting) is not an option here. We need a back-off strategy, and a way to `yield` if there are active fibers.